mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala

```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+    val unigramTable = generateUnigramTable(normalizedCumWeights, unigramTableSize)
+
+    (vocabMap.size, totalWordCount, vocabMap, unigramTable)
+  }
+
+  /**
+   * This method implements Word2Vec Continuous Bag Of Words based implementation using
+   * negative sampling optimization, using BLAS for vectorizing operations where applicable.
+   * The algorithm is parallelized in the same way as the skip-gram based estimation.
+   * @param input
+   * @return
+   */
+  private def fitCBOW[S <: Iterable[String]](input: RDD[S]): feature.Word2VecModel = {
+    val (vocabSize, totalWordCount, vocabMap, uniTable) = generateVocab(input)
+    val negSamples = $(negativeSamples)
+    assert(negSamples < vocabSize,
+      s"Vocab size ($vocabSize) cannot be smaller than negative samples($negSamples)")
+    val seed = $(this.seed)
+    val initRandom = new XORShiftRandom(seed)
+
+    val vectorSize = $(this.vectorSize)
+
+    val syn0Global = Array.fill(vocabSize * vectorSize)(initRandom.nextFloat - 0.5f)
+    val syn1Global = Array.fill(vocabSize * vectorSize)(0.0f)
+
+    val sc = input.context
+
+    val vocabMapbc = sc.broadcast(vocabMap)
+    val unigramTablebc = sc.broadcast(uniTable)
+
+    val window = $(windowSize)
+
+    val digitSentences = input.flatMap{sentence =>
+      val wordIndexes = sentence.flatMap(vocabMapbc.value.get)
+      wordIndexes.grouped($(maxSentenceLength)).map(_.toArray)
+    }.repartition($(numPartitions)).cache()
+
+    val learningRate = $(stepSize)
+
+    val wordsPerPartition = totalWordCount / $(numPartitions)
+
+    logInfo(s"VocabSize: ${vocabMap.size}, TotalWordCount: $totalWordCount")
+
+    for {iteration <- 1 to $(maxIter)} {
+      logInfo(s"Starting iteration: $iteration")
+      val iterationStartTime = System.nanoTime()
+
+      val syn0bc = sc.broadcast(syn0Global)
+      val syn1bc = sc.broadcast(syn1Global)
+
+      val partialFits = digitSentences.mapPartitionsWithIndex{ case (i_, iter) =>
+        logInfo(s"Iteration: $iteration, Partition: $i_")
+        logInfo(s"Numerical lib class being used : ${blas.getClass.getName}")
+        val random = new XORShiftRandom(seed ^ ((i_ + 1) << 16) ^ ((-iteration - 1) << 8))
+        val contextWordPairs = iter.flatMap(generateContextWordPairs(_, window, random))
+
+        val groupedBatches = contextWordPairs.grouped(batchSize)
+
+        val negLabels = 1.0f +: Array.fill(negSamples)(0.0f)
+        val syn0 = syn0bc.value
+        val syn1 = syn1bc.value
+        val unigramTable = unigramTablebc.value
+
+        // initialize intermediate arrays
+        val contextVector = Array.fill(vectorSize)(0.0f)
+        val l2Vectors = Array.fill(vectorSize * (negSamples + 1))(0.0f)
+        val gb = Array.fill(negSamples + 1)(0.0f)
+        val hiddenLayerUpdate = Array.fill(vectorSize * (negSamples + 1))(0.0f)
+        val neu1e = Array.fill(vectorSize)(0.0f)
+        val wordIndices = Array.fill(negSamples + 1)(0)
```

Comment:
Looks like your initializing all of these arrays to zero. Just do ``new Array[Float/Int](size)`` which is faster